node/pkg/vaa/structs.go

```diff
@@ -415,20 +425,100 @@ func (v *VAA) AddSignature(key *ecdsa.PrivateKey, index uint8) {
 	})
 }
 
+// NOTE: This function assumes that the caller has verified that the VAA is from the token bridge.
+func IsTransfer(payload []byte) bool {
+	if len(payload) > 0 {
+		if payload[0] == 1 || payload[0] == 3 {
+			return true
+		}
+	}
+
+	return false
+}
+
+func DecodeTransferPayloadHdr(payload []byte) (*TransferPayloadHdr, error) {
+	if !IsTransfer(payload) {
+		return nil, fmt.Errorf("unsupported payload type")
+	}
+
+	if len(payload) < 101 {
+		return nil, fmt.Errorf("buffer too short")
+	}
+
+	p := &TransferPayloadHdr{}
+
+	// Payload type: payload[0]
+	p.Type = uint8(payload[0])
+
+	// Amount: payload[1] for 32
+	p.Amount = new(big.Int)
+	p.Amount.SetBytes(payload[1:33])
+
+	reader := bytes.NewReader(payload[33:])
+
+	// Token address: payload[33] for 32
+	p.TokenAddress = Address{}
+	err := binary.Read(reader, binary.BigEndian, &p.TokenAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	// Token chain ID: payload[65] for 2
+	p.TokenChainID = ChainID(0)
+	err = binary.Read(reader, binary.BigEndian, &p.TokenChainID)
+	if err != nil {
+		return nil, err
+	}
+
+	// To address: payload[67] for 32
+	p.ToAddress = Address{}
+	err = binary.Read(reader, binary.BigEndian, &p.ToAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	// To chain ID: payload[99] for 2
+	p.ToChainID = ChainID(0)
+	err = binary.Read(reader, binary.BigEndian, &p.ToChainID)
+	if err != nil {
+		return nil, err
+	}
+
+	return p, nil
+}
+
 // MustWrite calls binary.Write and panics on errors
 func MustWrite(w io.Writer, order binary.ByteOrder, data interface{}) {
 	if err := binary.Write(w, order, data); err != nil {
 		panic(fmt.Errorf("failed to write binary data: %v", data).Error())
 	}
 }
 
-// StringToAddress converts a hex-encoded adress into a vaa.Address
+// StringToAddress converts a hex-encoded address into a vaa.Address
 func StringToAddress(value string) (Address, error) {
 	var address Address
+
+	// Make sure we have enough to decode
+	if len(value) < 2 {
+		return address, fmt.Errorf("value must be at least 1 byte")
+	}
+
+	// Trim any preceding "0x" to the address
+	if value[0:2] == "0x" {
+		value = value[2:]
+	}
```

Comment:
The previous logic was overloading a condition where you could supply a sub-1 byte hex string which would trip on `hex.DecodeString() ` step.

https://github.com/certusone/wormhole/blob/fa49e2a375a721a2715cd6f1e1c0de03291ab8be/node/pkg/vaa/structs.go#L493-L495